### PR TITLE
Turn on `xfail_strict = true` for all python packages

### LINKED ISSUE #12391: Turn on `xfail_strict=true` in all subpackages
Followup to #12244.

Can we do the same for other packages in the repo? I would expect that dask-cudf / custreamz / cudf-kafka / strings_udf won't have as many problems as cudf.

_Originally posted by @bdice in https://github.com/rapidsai/cudf/pull/12244#discussion_r1048803383_

### LINKED ISSUE #16974: [BUG] some pytest configurations not used
**Describe the bug**
Since #16851, all python packages have a `pytest.ini` in `package/tests/`. However, some of the packages previously had pytest-specific configuration in the `pytest.tool.ini-options` section of `pyproject.toml`. These settings are now ignored (as per https://docs.pytest.org/en/8.2.x/reference/customize.html#finding-the-rootdir).

We should pick a single configuration file that is used (and recommended in the dev guide).

FWIW, my preference is for pyproject.toml, since that's where all the other tool-specific config is going.

### PR DESCRIPTION
**Pull request by wence-**
## Description

The cudf tests already treat tests that are expected to fail but pass as errors, but at the time we introduced that change, we didn't do the same for the other packages. Do that now, it turns out there are only a few xpassing tests.

While here, it turns out that having multiple different pytest configuration files does not work. `pytest.ini` takes precedence over other options, and it's "first file wins". Consequently, the merge of #16851 turned off `xfail_strict = true` (and other options) for many of the subpackages.

To fix this, migrate all pytest configuration into the appropriate section of the `pyproject.toml` files, so that all tool configuration lives in the same place. We also add a section in the developer guide to document this choice.

- Closes #12391 
- Closes #16974

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Review by jameslamb — APPROVED** [review]
Totally support consolidating all these settings in `pyproject.toml`... that mix of `pytest.ini` and `pyproject.toml` here has tripped me up in the past too.

**Review by mroeschke — APPROVED** [review]
Some optional options to set in a follow-up if interested to make pytest more strict

* `empty_parameter_set_mark = "fail_at_collect"` : https://docs.pytest.org/en/stable/reference/reference.html#confval-empty_parameter_set_mark
* `addopts = "--strict-config --strict-markers"`

**Review by galipremsagar — APPROVED** [review]

**Comment by wence-** [comment]
> * --strict-config --strict-markers

I added these, let's see

File: python/cudf/pyproject.toml [hunk 1]
```diff
@@ -124,6 +124,27 @@ skip = [
...
+    "ignore:::.*xdist.*",
+    "ignore:::.*pytest.*",
+    # some third-party dependencies (e.g. 'boto3') still using datetime.datetime.utcnow()
+    "ignore:.*datetime.*utcnow.*scheduled for removal.*:DeprecationWarning:botocore",
+    # Deprecation warning from Pyarrow Table.to_pandas() with pandas-2.2+
+    "ignore:Passing a BlockManager to DataFrame is deprecated:DeprecationWarning",
+    # PerformanceWarning from cupy warming up the JIT cache
+    "ignore:Jitify is performing a one-time only warm-up to populate the persistent cache:cupy._util.PerformanceWarning",
+    # Ignore numba PEP 456 warning specific to arm machines
+    "ignore:FNV hashing is not implemented in Numba.*:UserWarning"
```
Comment (Matt711) [inline python/cudf/pyproject.toml:141]: Maybe?
```suggestion
    "ignore:.*FNV hashing is not implemented in Numba.*:UserWarning"
```

Comment (wence-) [reply python/cudf/pyproject.toml:141]: Thanks, I think it is that I also need to apply the same filter to the custreamz/cudf_kafka tests

**Pull request merged**